### ISSUE TITLE
GenProSeq

### ISSUE THREAD

**Issue by dongminjung** [Bioconductor/Contributions#2304]
Update the following URL to point to the GitHub repository of
the package you wish to submit to _Bioconductor_

- Repository: https://github.com/dongminjung/GenProSeq

Confirm the following by editing each check box to '[x]'

- [x] I understand that by submitting my package to _Bioconductor_,
  the package source and all review commentary are visible to the
  general public.

- [x] I have read the _Bioconductor_ [Package Submission][2]
  instructions. My package is consistent with the _Bioconductor_
  [Package Guidelines][1].

- [x] I understand Bioconductor [Package Naming Policy][9] and acknowledge
  Bioconductor may retain use of package name.

- [x] I understand that a minimum requirement for package acceptance 
  is to pass R CMD check and R CMD BiocCheck with no ERROR or WARNINGS. 
  Passing these checks does not result in automatic acceptance. The 
  package will then undergo a formal review and recommendations for 
  acceptance regarding other Bioconductor standards will be addressed.

- [x] My package addresses statistical or bioinformatic issues related
  to the analysis and comprehension of high throughput genomic data.

- [x] I am committed to the long-term maintenance of my package. This
  includes monitoring the [support site][3] for issues that users may
  have, subscribing to the [bioc-devel][4] mailing list to stay aware
  of developments in the _Bioconductor_ community, responding promptly
  to requests for updates from the Core team in response to changes in
  _R_ or underlying software.
  
- [x] I am familiar with the [Bioconductor code of conduct][7] and 
  agree to abide by it.

I am familiar with the essential aspects of _Bioconductor_ software
management, including:

- [x] The 'devel' branch for new packages and features.
- [x] The stable 'release' branch, made available every six
      months, for bug fixes.
- [x] _Bioconductor_ version control using [Git][5]
  (optionally [via GitHub][6]).

For questions/help about the submission process, including questions about
the output of the automatic reports generated by the SPB (Single Package
Builder), please use the #package-submission channel of our Community Slack.
Follow the link on the home page of the [Bioconductor website][8] to sign up.

[1]: https://contributions.bioconductor.org/
[2]: https://bioconductor.org/developers/package-submission/
[3]: https://support.bioconductor.org
[4]: https://stat.ethz.ch/mailman/listinfo/bioc-devel
[5]: http://bioconductor.org/developers/how-to/git/
[6]: http://bioconductor.org/developers/how-to/git/sync-existing-repositories/
[7]: https://bioconductor.org/about/code-of-conduct/
[8]: https://bioconductor.org/
[9]: https://bioconductor.org/developers/package-submission/#naming

**Comment by nturaga** [comment]
@dongminjung  Issue another build report please. We'll try to see if the Windows error is resolved on the build system already. 

Please note the release schedule that is posted here - http://bioconductor.org/developers/release-schedule/.

We'll try to get your package in before that.

**Comment by lshep** [comment]
We are closing this issue for inactivity.  If you would like to continue the review process please comment here for the issue to be reopened.

**Comment by lshep** [comment]
You were asked to kick off a new build and didn't.   also please see the october mailing list thread about using JAVA_HOME.  https://stat.ethz.ch/pipermail/bioc-devel/2021-October/018594.html

**Comment by dongminjung** [comment]
The issue for JAVA_HOME is resolved by adding `.BBSoptions`. However, there is another issue. Please, find that. Thanks.

**Comment by hpages** [comment]
Hi @dongminjung,

`GenProSeq::fit_ART()` calls `DeepPINCS::get_seq_encode_pad()` internally which does not seem to work properly on Windows.

In a _fresh_ R session on riesling1:
```
library(DeepPINCS)
get_seq_encode_pad("MSPLNQSAEGLPQEASNRSLNATETSEAW", 10)  # 1st attempt fails!
# 2021-11-30 02:21:00.436544: W tensorflow/stream_executor/platform/default/dso_loader.cc:59] Could not load dynamic library 'cudart64_101.dll'; dlerror: cudart64_101.dll not found
# 2021-11-30 02:21:00.438851: I tensorflow/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
# Loaded Tensorflow version 2.3.1
# Error:
```
Unfortunately the error message is empty.

Interestingly, the 2nd attempt works:
```
get_seq_encode_pad("MSPLNQSAEGLPQEASNRSLNATETSEAW", 10)  # 2nd attempt works!
# $sequences_encode_pad
#      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
# [1,]    4    6    1   11    2   11   10    2    1    12
#
# $lenc
# An object of class "LabelEncoder.Character"
# Slot "classes":
#  [1] "A" "E" "G" "L" "M" "N" "P" "Q" "R" "S" "T" "W"
#
# Slot "type":
# [1] "character"
#
# Slot "mapping":
#   classes ind
# 1        A   1
# 2        E   2
# 3        G   3
# 4        L   4
# 5        M   5
# 6        N   6
# 7        P   7
# 8        Q   8
# 9        R   9
# 10       S  10
# 11       T  11
# 12       W  12
# 
# 
# $num_tokens
# [1] 12
```
Furthermore, it works on the first attempt if I explicitly configure Python before calling `get_seq_encode_pad()`. For example this works in a _fresh_ R session:
```
reticulate::py_config()
library(DeepPINCS)
get_seq_encode_pad("MSPLNQSAEGLPQEASNRSLNATETSEAW", 10)  # 1st attempt works!
```
So there seems to be some Python initialization issue that affects **DeepPINCS** on Windows.

Hard to know what the good practices are w.r.t. Python initialization. Is it the responsibility of the end user to call `reticulate::py_config()` before using a **reticulate**-based package? Or should each **reticulate**-based function take care of this for the end-user (e.g. `get_seq_encode_pad()` calls `reticulate::py_config()` before doing anything else).

Unfortunately trying to find some guidance by looking at what the **reticulate** experts do in CRAN packages **tensorflow** or **keras** does not help because these packages completely avoid the problem by providing zero running examples in their man pages, zero evaluated code chunks in their vignettes, and by disabling all their unit tests by default. *sigh*

Is this something where the use of **basilisk** would help? IIRC you explored its use when you submitted **DeepPINCS** earlier this year but in the end it seems that you've dismissed it. Did you run into blocking issues?

Best,
H.

**Comment by dongminjung** [comment]
Hi @hpages

Thank you for your concerns about the python connection. Instead of **basilisk**, I add a condition for the connection to running examples by using `reticulate::py_module_available("rpytools")`. The **basilisk** uses a conda environment, however, I don't want to specify any versions of python packages to keep up with changes. Anyway, I found another error in `DESCRIPTION`.  I think, one possible reason is the different version of BiocCheckGitClone.

**Comment by hpages** [comment]
The problem is that for some reason `reticulate::py_module_available("rpytools")` always seems to return FALSE on Windows after loading **DeepPINCS**. This is on riesling1:

![Screenshot from 2021-12-02 04-31-26](https://user-images.githubusercontent.com/8810451/144423264-62139caa-6eb4-47bb-bf28-cec5a259f388.png)

No such problem if **DeepPINCS** is _not_ loaded before calling `reticulate::py_module_available("rpytools")`:

![Screenshot from 2021-12-02 04-33-31](https://user-images.githubusercontent.com/8810451/144423383-bdd74144-0e5a-42ee-b59d-76ddca452fa8.png)

Note that we don't see this with your other package **ttgsea** which also depends on **reticulate**, **tensorflow**, and **keras**:

![Screenshot from 2021-12-02 04-55-33](https://user-images.githubusercontent.com/8810451/144426733-c70dbb97-7470-4487-bb2a-3e789d199055.png)

This is why the `DeepPINCS::get_seq_encode_pad()` example _seems_ to work on Windows, even though the function doesn't work. **DeepPINCS** is all green on the daily build report only because the example in the man page for `get_seq_encode_pad()` doesn't actually call the function on Windows.

In other words, testing for `reticulate::py_module_available("rpytools")` before running the examples doesn't solve the problem, it only masks it.

This is something that needs to be addressed. You will probably need access to a Windows machine to sort this out.

Please remove all the calls to `reticulate::py_module_available("rpytools")` from packages **DeepPINCS**, **VAExprs**, and **GenProSeq**, so we get build reports for these packages that actually reflect the truth on Windows.

If for some reason Windows cannot be supported, we'll mark these packages as unsupported on this platform. 

Thanks,
H.

**Comment by dongminjung** [comment]
Hi @hpages

I found that the package **rcdk** causes the problem for Windows. So I removed it from the `NAMESPACE` file of **DeepPINCS**. Also, `reticulate::py_module_available("rpytools")` was removed from the packages, **DeepPINCS**, **VAExprs** and **GenProSeq**. Please find them. Thanks.

**Comment by nturaga** [comment]
The default branch in this repo is `main`. It will be copied as `master`, consider just using `master` as your default branch. 

https://github.com/github/renaming

**Comment by nturaga** [comment]
# Review

Build report is mostly alright but there are some NOTE's that you can easily fix. Please follow http://contributions.bioconductor.org/r-code.html#coding-style


## DESCRIPTION

ok

## NAMESPACE 

ok

## NEWS

ok 

## R 

- In your functions, check arguments to make sure valid arguments are
  being passed. Make the function fault-tolerant to user inputs. 

## inst

## man

ok

## tests 

ok

## vignettes 

ok

**Comment by nturaga** [comment]
Issue another build report, but the package looks fine for the most part.

**Comment by dongminjung** [comment]
Hi @nturaga

I appreciate you taking the time to review the package.

- The default branch is changed to `master`.
- Some NOTEs that can be easily fixed are removed.
    - Update R version dependency from 3.5.0 to 4.2.0.
    - Avoid using '=' for assignment and use '<-' instead
- For `fit_ART`, `gen_ART`, `fit_VAE`, `gen_VAE`, `fit_GAN` and `gen_GAN`, arguments are checked for validity. Furthermore, to make fault-tolerant to user inputs, the error handling is used for invalid inputs by `tryCatch`.

Please let me know if there is anything necessary to be modified. Thanks.

**Comment by nturaga** [comment]
Hi, @dongminjung 

Your package looks good, but has warnings in the build report. Please solve them. Issue a new build report as well with a version bump.

I also wonder how you are writing the man packages without using 'roxygen2' when your NAMESPACE suggests you are using 'roxygen2.

**Comment by nturaga** [comment]
Please also include the version of Tensorflow and Keras and python you are using in a requirements.txt file included in the `inst/extdata` just for reference sake. We discussed this on a separate thread for your other package 'DeepPINCS'. I believe the same issues might pop up here. 

I was wondering if you tried to use `basilisk` (http://bioconductor.org/packages/release/bioc/html/basilisk.html)  for your package.

**Comment by dongminjung** [comment]
Hi @nturaga

I updated my code to meet your requirements as much as possible.

- In the build report, there are no warnings.
- In the NAMESPACE, 'roxygen2' was removed.
- In the requirements.txt, the versions of the python packages for tensorflow and keras are included. Also, README was updated.

**Comment by lshep** [comment]
The master branch of your GitHub repository has been added to Bioconductor's git repository.

To use the git.bioconductor.org repository, we need an 'ssh' key to associate with your github user name. If your GitHub account already has ssh public keys (https://github.com/dongminjung.keys is not empty), then no further steps are required. Otherwise, do the following:

1. [Add an SSH key to your github account][]
1. [Submit your SSH key to Bioconductor][]

See further instructions at

https://bioconductor.org/developers/how-to/git/

for working with this repository. See especially

https://bioconductor.org/developers/how-to/git/new-package-workflow/
https://bioconductor.org/developers/how-to/git/sync-existing-repositories/

to keep your GitHub and Bioconductor repositories in sync.

Your package will be included in the next nigthly 'devel' build (check-out from git at about 6 pm Eastern; build completion around 2pm Eastern the next day) at

https://bioconductor.org/checkResults/

(Builds sometimes fail, so ensure that the date stamps on the main landing page are consistent with the addition of your package). Once the package builds successfully, you package will be available for download in the 'Devel' version of Bioconductor using `BiocManager::install("GenProSeq")`. The package 'landing page' will be created at

https://bioconductor.org/packages/GenProSeq

If you have any questions, please contact the bioc-devel mailing list (https://stat.ethz.ch/mailman/listinfo/bioc-devel); this issue will not be monitored further.

[Add an SSH key to your github account]: https://help.github.com/articles/connecting-to-github-with-ssh/
[Submit your SSH key to Bioconductor]: https://git.bioconductor.org/BiocCredentials/
